src/Storages/StorageMerge.cpp

```diff
@@ -565,19 +670,19 @@ void registerStorageMerge(StorageFactory & factory)
                 " - name of source database and regexp for table names.",
                 ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
 
-        engine_args[0] = evaluateConstantExpressionForDatabaseName(engine_args[0], args.getLocalContext());
-        engine_args[1] = evaluateConstantExpressionAsLiteral(engine_args[1], args.getLocalContext());
+        auto [is_regexp, source_database_name_or_regexp] = evaluateDatabaseNameForMergeEngine(engine_args[0], args.getLocalContext());
 
-        String source_database = engine_args[0]->as<ASTLiteral &>().value.safeGet<String>();
+        engine_args[1] = evaluateConstantExpressionAsLiteral(engine_args[1], args.getLocalContext());
         String table_name_regexp = engine_args[1]->as<ASTLiteral &>().value.safeGet<String>();
```

Comment:
Looks like this removing of assignment to `engine_args[0]` is a reason of incorrect `show create table query` from failed test:
```code
 -CREATE TABLE default.merge\n(\n    `CounterID` UInt32,\n    `dummy` String,\n    `StartDate` Date,\n    `Sign` Int8,\n    `VisitID` UInt64,\n    `UserID` UInt64,\n    `StartTime` DateTime,\n    `ClickLogID` UInt64\n)\nENGINE = Merge(\'default\', \'merge\\\\[0-9\\\\]\')
 +CREATE TABLE default.merge\n(\n    `CounterID` UInt32,\n    `dummy` String,\n    `StartDate` Date,\n    `Sign` Int8,\n    `VisitID` UInt64,\n    `UserID` UInt64,\n    `StartTime` DateTime,\n    `ClickLogID` UInt64\n)\nENGINE = Merge(currentDatabase(), \'merge\\\\[0-9\\\\]\')
```